### PR TITLE
Backup schema optimization report in end migration, rename skip-recommendations flag

### PR DESCRIPTION
**Pull request by priyanshi-yb**
### Describe the changes in this pull request
1. Backup schema optimization report in the end of migration #2890 
2. Rename the `skip-recommendations` flag to `skip-colocation-recommendations` and `skip-performance-optimizations` to `skip-performance-recommendations` #2956 
3. Generating a JSON schema optimization report for testing purposes, not mentioning it to users
4. Added end-to-end tests for export and import with/without skip-performance optimizations.
5. Minor changes to the schema optimization report for some edge cases
    - Do not add redundant index changes in schema optimization if no redundant indexes are present in the schema
    - Consider Secondary Index to Range sharded change in `HasOptimizations` to generate a report if only this change is made in the schema

### Describe if there are any user-facing changes

Yes

### How was this pull request tested?

Added unit tests and manually

### Does your PR have changes that can cause upgrade issues? 
| Component        | Breaking changes? |
| :----------------------------------------------: | :-----------: |
| MetaDB                                                    | No |
| Name registry json                                        | No |
| Data File Descriptor Json                                 | No |
| Export Snapshot Status Json                               | No |
| Import Data State                                         |  No |
| Export Status Json                                        | No |
| Data .sql files of tables                                 | No |
| Export and import data queue                              | No |
| Schema Dump                                               | No |
| AssessmentDB                                              | No |
| Sizing DB                                                 | No |
| Migration Assessment Report Json                          | No |
| Callhome Json                                             | No |
| YugabyteD Tables                                          | No |
| TargetDB Metadata Tables                                  | No |

### CODE REVIEW

File: yb-voyager/cmd/templates/schema_optimization_report.template [hunk 1]
```diff
@@ -515,7 +515,7 @@
                </p>
                <ul style="margin: 0.5rem 0 0 0; padding-left: 1.5rem; color: #666; font-size: 0.9rem;">
                   <li><strong>--skip-performance-optimizations true</strong> - Skip performance optimizations (e.g. redundant index removal)</li>
-                  <li><strong>--skip-recommendations true</strong> - Skip sharding recommendations from assessment</li>
+                  <li><strong>--skip-sharding-recommendations true</strong> - Skip sharding recommendations from assessment</li>
```
Comment (makalaaneesh) [inline yb-voyager/cmd/templates/schema_optimization_report.template:518]: this could be interpreted as "range vs hash" decision. Let's think a little more about the name. Perhaps "skip-colocation-recommendations"?

File: yb-voyager/cmd/endMigrationCommand.go [hunk 2]
```diff
@@ -286,6 +287,36 @@ func saveMigrationReportsFn(msr *metadb.MigrationStatusRecord) {
...
+	alreadyBackedUp := utils.FileOrFolderExistsWithGlobPattern(optimizationReportGlobPath)
+	if alreadyBackedUp {
+		utils.PrintAndLog("schema optimization report is already present at %q", optimizationReportGlobPath)
+		return
+	}
+	utils.PrintAndLog("saving schema optimization report...")
+
+	files, err := os.ReadDir(filepath.Join(exportDir, "reports"))
+	if err != nil {
+		utils.ErrExit("reading reports directory: %w", err)
```
Comment (makalaaneesh) [inline yb-voyager/cmd/endMigrationCommand.go:301]: let's return errors properly.

File: yb-voyager/cmd/endMigrationCommand.go [hunk 3]
```diff
@@ -286,6 +287,36 @@ func saveMigrationReportsFn(msr *metadb.MigrationStatusRecord) {
...
+	}
+	for _, file := range files {
```
Comment (makalaaneesh) [inline yb-voyager/cmd/endMigrationCommand.go:303]: we know the exact file location. Can we not directly check if it exists or not. Why get the files in the reports, and then do a prefix match?

File: yb-voyager/cmd/exportSchema_test.go [hunk 4]
```diff
@@ -335,3 +338,229 @@ func TestExportSchemaRunningAssessmentInternally_DisableFlag(t *testing.T) {
 		t.Errorf("Expected assessment report file does exist: %s", reportFilePath)
 	}
 }
+
+//Test: export schema default behavior
+//Expectation: schema optimization report should be generated with performance optimizations applied
+//Run import schema to verify that indexes are changed/removed
+//verify the indexes on YB
+func TestExportSchemaSchemaOptimizationReportPerfOptimizationsAutofix(t *testing.T) {
```
Comment (makalaaneesh) [inline yb-voyager/cmd/exportSchema_test.go:346]: nice!
